File: docs/chapter2/transformer-python/transformer_pytorch/core/encoder.py
"""
编码器模块

实现 Transformer 编码器层和编码器块：
- 编码器层 (EncoderLayer)
- Transformer 编码器 (TransformerEncoder)

编码器层包含多头自注意力和前馈网络，以及残差连接和层归一化。

作者: Transformer-PyTorch Team
版本: 1.0.0
"""

from typing import List, Optional, Tuple

import torch
import torch.nn as nn
from torch import Tensor

from .attention import MultiHeadAttention, SelfAttention
from .layers import FeedForward, LayerNorm, ResidualConnection
from .math_utils import init_weights


class EncoderLayer(nn.Module):
    """
    Transformer 编码器层
    
    实现单个编码器层，包含：
    1. 多头自注意力机制
    2. 残差连接和层归一化
    3. 位置相关的前馈网络
    4. 残差连接和层归一化
    
    Args:
        d_model (int): 模型维度
        num_heads (int): 注意力头数
        d_ff (int): 前馈网络隐藏层维度
        dropout (float): Dropout 概率，默认为 0.1
        activation (str): 激活函数类型，默认为 'relu'
        norm_first (bool): 是否使用 Pre-LN 结构，默认为 True
        
    Examples:
        >>> layer = EncoderLayer(d_model=512, num_heads=8, d_ff=2048)
        >>> x = torch.randn(2, 10, 512)
        >>> output, attention = layer(x)
        >>> print(output.shape)  # torch.Size([2, 10, 512])
    """
    
    def __init__(
        self,
        d_model: int,
        num_heads: int,
        d_ff: int,
        dropout: float = 0.1,
        activation: str = 'relu',
        norm_first: bool = True
    ):
        super().__init__()
        self.d_model = d_model
        self.num_heads = num_heads
        self.d_ff = d_ff
        self.norm_first = norm_first
        
        # 多头自注意力
        self.self_attention = SelfAttention(
            d_model=d_model,
            num_heads=num_heads,
            dropout=dropout
        )
        
        # 前馈网络
        self.feed_forward = FeedForward(
            d_model=d_model,
            d_ff=d_ff,
            activation=activation,
            dropout=dropout
        )
        
        # 层归一化
        self.norm1 = LayerNorm(d_model)
        self.norm2 = LayerNorm(d_model)
        
        # Dropout
        self.dropout1 = nn.Dropout(dropout)
        self.dropout2 = nn.Dropout(dropout)
        
        # 初始化权重
        self.apply(lambda module: init_weights(module, init_std=0.02))
    
    def forward(
        self,
        x: Tensor,
        mask: Optional[Tensor] = None,
        return_attention: bool = False
    ) -> Tuple[Tensor, Optional[Tensor]]:
        """
        前向传播
        
        Args:
            x (Tensor): 输入张量 [batch_size, seq_len, d_model]
            mask (Optional[Tensor]): 注意力掩码 [batch_size, seq_len, seq_len]
            return_attention (bool): 是否返回注意力权重
            
        Returns:
            Tuple[Tensor, Optional[Tensor]]: (输出张量, 注意力权重)
        """
        if self.norm_first:
            # Pre-LN 结构: LayerNorm -> SubLayer -> Dropout -> Residual
            
            # 第一个子层：自注意力
            norm_x = self.norm1(x)
            attn_output, attention_weights = self.self_attention(
                norm_x, mask=mask, return_attention=True
            )
            x = x + self.dropout1(attn_output)
            
            # 第二个子层：前馈网络
            norm_x = self.norm2(x)
            ff_output = self.feed_forward(norm_x)
            x = x + self.dropout2(ff_output)
            
        else:
            # Post-LN 结构: SubLayer -> Dropout -> Residual -> LayerNorm
            
            # 第一个子层：自注意力
            attn_output, attention_weights = self.self_attention(
                x, mask=mask, return_attention=True
            )
            x = self.norm1(x + self.dropout1(attn_output))
            
            # 第二个子层：前馈网络
            ff_output = self.feed_forward(x)
            x = self.norm2(x + self.dropout2(ff_output))
        
        if return_attention:
            return x, attention_weights
        else:
            return x, None
    
    def extra_repr(self) -> str:
        return f'd_model={self.d_model}, num_heads={self.num_heads}, d_ff={self.d_ff}, norm_first={self.norm_first}'


class TransformerEncoder(nn.Module):
    """
    Transformer 编码器
    
    由多个编码器层堆叠而成，可选择性地在最后添加层归一化。
    
    Args:
        encoder_layer (EncoderLayer): 编码器层实例
        num_layers (int): 编码器层数
        norm (Optional[nn.Module]): 最终的归一化层，默认为 None
        
    Examples:
        >>> encoder_layer = EncoderLayer(d_model=512, num_heads=8, d_ff=2048)
        >>> encoder = TransformerEncoder(encoder_layer, num_layers=6)
        >>> x = torch.randn(2, 10, 512)
        >>> output = encoder(x)
        >>> print(output.shape)  # torch.Size([2, 10, 512])
    """
    
    def __init__(
        self,
        encoder_layer: EncoderLayer,
        num_layers: int,
        norm: Optional[nn.Module] = None
    ):
        super().__init__()
        self.num_layers = num_layers
        
        # 复制编码器层
        self.layers = nn.ModuleList([
            self._get_cloned_layer(encoder_layer) for _ in range(num_layers)
        ])
        
        # 最终的层归一化
        self.norm = norm
    
    def _get_cloned_layer(self, layer: nn.Module) -> nn.Module:
        """克隆编码器层"""
        import copy
        return copy.deepcopy(layer)
    
    def forward(
        self,
        x: Tensor,
        mask: Optional[Tensor] = None,
        return_all_layers: bool = False,
        return_attention: bool = False
    ) -> Tensor:
        """
        前向传播
        
        Args:
            x (Tensor): 输入张量 [batch_size, seq_len, d_model]
            mask (Optional[Tensor]): 注意力掩码
            return_all_layers (bool): 是否返回所有层的输出
            return_attention (bool): 是否返回注意力权重
            
        Returns:
            Tensor: 编码器输出 [batch_size, seq_len, d_model]
            或者包含所有层输出和注意力权重的字典
        """
        output = x
        all_layer_outputs = []
        all_attention_weights = []
        
        # 逐层前向传播
        for layer in self.layers:
            output, attention_weights = layer(
                output, mask=mask, return_attention=return_attention
            )
            
            if return_all_layers:
                all_layer_outputs.append(output)
            
            if return_attention and attention_weights is not None:
                all_attention_weights.append(attention_weights)
        
        # 最终的层归一化
        if self.norm is not None:
            output = self.norm(output)
        
        # 根据返回选项组织输出
        if return_all_layers or return_attention:
            result = {'last_hidden_state': output}
            
            if return_all_layers:
                result['all_hidden_states'] = all_layer_outputs
            
            if return_attention:
                result['all_attention_weights'] = all_attention_weights
            
            return result
        else:
            return output
    
    def get_layer_output(
        self,
        x: Tensor,
        layer_idx: int,
        mask: Optional[Tensor] = None
    ) -> Tensor:
        """
        获取指定层的输出
        
        Args:
            x (Tensor): 输入张量
            layer_idx (int): 层索引（0-based）
            mask (Optional[Tensor]): 注意力掩码
            
        Returns:
            Tensor: 指定层的输出
        """
        if layer_idx < 0 or layer_idx >= self.num_layers:
            raise ValueError(f"层索引 {layer_idx} 超出范围 [0, {self.num_layers - 1}]")
        
        output = x
        for i, layer in enumerate(self.layers):
            output, _ = layer(output, mask=mask, return_attention=False)
            if i == layer_idx:
                return output
        
        return output


def create_encoder(
    d_model: int,
    num_heads: int,
    d_ff: int,
    num_layers: int,
    dropout: float = 0.1,
    activation: str = 'relu',
    norm_first: bool = True,
    final_norm: bool = True
) -> TransformerEncoder:
    """
    创建 Transformer 编码器的便捷函数
    
    Args:
        d_model (int): 模型维度
        num_heads (int): 注意力头数
        d_ff (int): 前馈网络隐藏层维度
        num_layers (int): 编码器层数
        dropout (float): Dropout 概率
        activation (str): 激活函数类型
        norm_first (bool): 是否使用 Pre-LN 结构
        final_norm (bool): 是否在最后添加层归一化
        
    Returns:
        TransformerEncoder: 编码器实例
        
    Examples:
        >>> encoder = create_encoder(
        ...     d_model=512, num_heads=8, d_ff=2048, num_layers=6
        ... )
        >>> x = torch.randn(2, 10, 512)
        >>> output = encoder(x)
        >>> print(output.shape)  # torch.Size([2, 10, 512])
    """
    # 创建编码器层
    encoder_layer = EncoderLayer(
        d_model=d_model,
        num_heads=num_heads,
        d_ff=d_ff,
        dropout=dropout,
        activation=activation,
        norm_first=norm_first
    )
    
    # 创建最终的层归一化（如果需要）
    norm = LayerNorm(d_model) if final_norm else None
    
    # 创建编码器
    encoder = TransformerEncoder(
        encoder_layer=encoder_layer,
        num_layers=num_layers,
        norm=norm
    )
    
    return encoder


# 导出的类和函数列表
__all__ = [
    'EncoderLayer',
    'TransformerEncoder',
    'create_encoder',
]

File: docs/chapter2/transformer-js/core/encoder.js
/**
 * 编码器模块
 * 实现 Transformer 编码器层和编码器块
 * 
 * @author Transformer-JS
 * @version 1.0.0
 */

const { MultiHeadAttention } = require('./attention');
const { LayerNorm, MLP, ResidualConnection } = require('./layers');
const { add } = require('./matrix-ops');

/**
 * 编码器层
 * 实现单个 Transformer 编码器层
 * 包含多头自注意力和前馈网络，以及残差连接和层归一化
 */
class EncoderLayer {
  /**
   * 构造函数
   * 
   * @param {Object} config - 配置对象
   * @param {number} config.d_model - 模型维度
   * @param {number} config.n_heads - 注意力头数
   * @param {number} config.d_ff - 前馈网络隐藏层维度
   * @param {number} config.dropout - Dropout 概率
   * @param {string} config.activation - 激活函数类型
   * @param {boolean} config.use_bias - 是否使用偏置
   */
  constructor(config) {
    this.dModel = config.d_model || config.n_embd;
    this.nHeads = config.n_heads;
    this.dFF = config.d_ff || config.ffn_hidden_dim;
    this.dropout = config.dropout || 0.1;
    this.activation = config.activation || 'relu';
    this.useBias = config.use_bias || false;
    
    // 多头自注意力层
    this.selfAttention = new MultiHeadAttention(
      this.dModel,
      this.nHeads,
      this.dropout,
      this.useBias
    );
    
    // 前馈网络
    this.feedForward = new MLP(
      this.dModel,
      this.dFF,
      this.dModel,
      this.activation,
      this.dropout,
      this.useBias
    );
    
    // 层归一化
    this.norm1 = new LayerNorm(this.dModel);
    this.norm2 = new LayerNorm(this.dModel);
    
    // 残差连接
    this.residual1 = new ResidualConnection(this.dModel, this.dropout);
    this.residual2 = new ResidualConnection(this.dModel, this.dropout);
  }
  
  /**
   * 前向传播
   * 
   * @param {Array<Array<number>>} x - 输入矩阵 [seqLen, dModel]
   * @param {Array<Array<number>>|null} mask - 注意力掩码，可选
   * @returns {Object} {output: 编码器层输出, attention: 注意力权重}
   */
  forward(x, mask = null) {
    if (!Array.isArray(x) || x.length === 0) {
      throw new Error('输入必须是非空二维数组');
    }
    
    // 第一个子层：多头自注意力 + 残差连接 + 层归一化
    // Pre-LN 结构: x + attention(norm(x))
    const norm1Output = this.norm1.forward(x);
    const attentionResult = this.selfAttention.forward(norm1Output, norm1Output, norm1Output, mask);
    const afterAttention = add(x, attentionResult.output);
    
    // 第二个子层：前馈网络 + 残差连接 + 层归一化
    // Pre-LN 结构: x + ffn(norm(x))
    const norm2Output = this.norm2.forward(afterAttention);
    const ffnOutput = this.feedForward.forward(norm2Output);
    const output = add(afterAttention, ffnOutput);
    
    return {
      output: output,
      attention: attentionResult.attention
    };
  }
  
  /**
   * 设置训练模式
   * 
   * @param {boolean} training - 是否为训练模式
   */
  setTraining(training) {
    this.selfAttention.setTraining(training);
    this.feedForward.setTraining(training);
    this.residual1.setTraining(training);
    this.residual2.setTraining(training);
  }
  
  /**
   * 获取参数数量
   * 
   * @returns {number} 参数总数
   */
  getParameterCount() {
    return (
      this.selfAttention.getParameterCount() +
      this.feedForward.getParameterCount() +
      this.norm1.getParameterCount() +
      this.norm2.getParameterCount()
    );
  }
}

/**
 * Transformer 编码器
 * 由多个编码器层堆叠而成
 */
class TransformerEncoder {
  /**
   * 构造函数
   * 
   * @param {Object} config - 配置对象
   * @param {number} config.n_layers - 编码器层数
   * @param {number} config.d_model - 模型维度
   * @param {number} config.n_heads - 注意力头数
   * @param {number} config.d_ff - 前馈网络隐藏层维度
   * @param {number} config.dropout - Dropout 概率
   * @param {string} config.activation - 激活函数类型
   * @param {boolean} config.use_bias - 是否使用偏置
   */
  constructor(config) {
    this.nLayers = config.n_layers;
    this.dModel = config.d_model || config.n_embd;
    
    // 创建多个编码器层
    this.layers = [];
    for (let i = 0; i < this.nLayers; i++) {
      this.layers.push(new EncoderLayer(config));
    }
    
    // 最终的层归一化
    this.finalNorm = new LayerNorm(this.dModel);
  }
  
  /**
   * 前向传播
   * 
   * @param {Array<Array<number>>} x - 输入矩阵 [seqLen, dModel]
   * @param {Array<Array<number>>|null} mask - 注意力掩码，可选
   * @returns {Object} {output: 编码器输出, attentions: 各层注意力权重}
   */
  forward(x, mask = null) {
    if (!Array.isArray(x) || x.length === 0) {
      throw new Error('输入必须是非空二维数组');
    }
    
    let output = x;
    const attentions = [];
    
    // 逐层前向传播
    for (let i = 0; i < this.nLayers; i++) {
      const layerResult = this.layers[i].forward(output, mask);
      output = layerResult.output;
      attentions.push(layerResult.attention);
    }
    
    // 最终层归一化
    output = this.finalNorm.forward(output);
    
    return {
      output: output,
      attentions: attentions
    };
  }
  
  /**
   * 设置训练模式
   * 
   * @param {boolean} training - 是否为训练模式
   */
  setTraining(training) {
    for (const layer of this.layers) {
      layer.setTraining(training);
    }
  }
  
  /**
   * 获取参数数量
   * 
   * @returns {number} 参数总数
   */
  getParameterCount() {
    let totalParams = 0;
    
    // 各编码器层的参数
    for (const layer of this.layers) {
      totalParams += layer.getParameterCount();
    }
    
    // 最终层归一化的参数
    totalParams += this.finalNorm.getParameterCount();
    
    return totalParams;
  }
  
  /**
   * 获取指定层的输出
   * 用于分析中间层的表示
   * 
   * @param {Array<Array<number>>} x - 输入矩阵
   * @param {number} layerIndex - 层索引（0-based）
   * @param {Array<Array<number>>|null} mask - 注意力掩码，可选
   * @returns {Object} 指定层的输出和注意力权重
   */
  getLayerOutput(x, layerIndex, mask = null) {
    if (layerIndex < 0 || layerIndex >= this.nLayers) {
      throw new Error(`层索引 ${layerIndex} 超出范围 [0, ${this.nLayers - 1}]`);
    }
    
    let output = x;
    
    // 前向传播到指定层
    for (let i = 0; i <= layerIndex; i++) {
      const layerResult = this.layers[i].forward(output, mask);
      output = layerResult.output;
      
      if (i === layerIndex) {
        return {
          output: output,
          attention: layerResult.attention
        };
      }
    }
  }
  
  /**
   * 获取所有层的输出
   * 用于分析模型的层级表示
   * 
   * @param {Array<Array<number>>} x - 输入矩阵
   * @param {Array<Array<number>>|null} mask - 注意力掩码，可选
   * @returns {Array<Object>} 各层的输出和注意力权重
   */
  getAllLayerOutputs(x, mask = null) {
    let output = x;
    const layerOutputs = [];
    
    for (let i = 0; i < this.nLayers; i++) {
      const layerResult = this.layers[i].forward(output, mask);
      output = layerResult.output;
      
      layerOutputs.push({
        layerIndex: i,
        output: output,
        attention: layerResult.attention
      });
    }
    
    // 添加最终层归一化后的输出
    const finalOutput = this.finalNorm.forward(output);
    layerOutputs.push({
      layerIndex: 'final',
      output: finalOutput,
      attention: null
    });
    
    return layerOutputs;
  }
}

/**
 * 编码器工厂函数
 * 根据配置创建编码器实例
 * 
 * @param {Object} config - 配置对象
 * @returns {TransformerEncoder} 编码器实例
 */
function createEncoder(config) {
  // 验证必需的配置参数
  const requiredParams = ['n_layers', 'n_heads'];
  for (const param of requiredParams) {
    if (!(param in config)) {
      throw new Error(`缺少必需的配置参数: ${param}`);
    }
  }
  
  // 设置默认值
  const defaultConfig = {
    d_model: config.n_embd || 512,
    d_ff: config.ffn_hidden_dim || (config.n_embd || 512) * 4,
    dropout: 0.1,
    activation: 'relu',
    use_bias: false
  };
  
  const mergedConfig = { ...defaultConfig, ...config };
  
  return new TransformerEncoder(mergedConfig);
}

// 导出类和函数
module.exports = {
  EncoderLayer,
  TransformerEncoder,
  createEncoder
};
